core/src/main/java/org/jboss/pnc/build/finder/pnc/client/CachedPncClient.java

```diff
@@ -18,77 +18,103 @@
 import org.jboss.pnc.client.RemoteCollection;
 import org.jboss.pnc.client.RemoteResourceException;
 import org.jboss.pnc.dto.Artifact;
 import org.jboss.pnc.dto.BuildPushResult;
 import org.jboss.pnc.dto.ProductVersion;
 
 /**
- * Implementation of adapter to communicate with PNC Orchestrator REST API, which caches the results in HashMaps to
- * improve performance of the application
+ * Implementation of adapter to communicate with PNC Orchestrator REST API, which caches the results in HashMaps and
+ * ISPN to improve performance of the application
  *
  * @author Jakub Bartecek
  */
-public class HashMapCachingPncClient implements PncClient {
-    private final PncClient pncClient;
-
-    private final Map<String, StaticRemoteCollection<Artifact>> md5Cache = new HashMap<>();
+public class CachedPncClient implements PncClient {
 
-    private final Map<String, StaticRemoteCollection<Artifact>> sha1Cache = new HashMap<>();
+    private final PncClient pncClient;
 
-    private final Map<String, StaticRemoteCollection<Artifact>> sha256Cache = new HashMap<>();
+    private final Map<String, ArtifactStaticRemoteCollection> artifactCache;
 
     private final Map<String, BuildPushResult> getBuildPushResultCache = new HashMap<>();
 
     private final Map<String, ProductVersion> getProductVersionCache = new HashMap<>();
 
-    public HashMapCachingPncClient(BuildConfig config) {
+    public CachedPncClient(BuildConfig config, BasicCacheContainer cacheManager) {
+        if (cacheManager == null) {
+            artifactCache = new HashMap<>();
+        } else {
+            artifactCache = cacheManager.getCache("artifact-pnc");
+        }
         this.pncClient = new PncClientImpl(config);
     }
 
-    public HashMapCachingPncClient(PncClient pncClient) {
+    public CachedPncClient(PncClient pncClient, BasicCacheContainer cacheManager) {
+        if (cacheManager == null) {
+            artifactCache = new HashMap<>();
+        } else {
+            artifactCache = cacheManager.getCache("artifact-pnc");
+        }
         this.pncClient = pncClient;
     }
 
     @Override
     public RemoteCollection<Artifact> getArtifactsByMd5(String md5) throws RemoteResourceException {
-        RemoteCollection<Artifact> cachedEntity = md5Cache.get(md5);
-        if (cachedEntity != null) {
-            return cachedEntity;
-        } else {
-            RemoteCollection<Artifact> foundEntity = pncClient.getArtifactsByMd5(md5);
-            StaticRemoteCollection<Artifact> staticRemoteCollection = new StaticRemoteCollection<>(foundEntity);
-            md5Cache.put(md5, staticRemoteCollection);
-            return staticRemoteCollection;
+        ArtifactStaticRemoteCollection cachedValue = getFromCache(md5);
+        if (cachedValue != null) {
+            return cachedValue;
+        }
+
+        RemoteCollection<Artifact> artifacts = pncClient.getArtifactsByMd5(md5);
+        if (artifacts != null && artifacts.size() > 0) {
```

Comment:
No, it does not... The `RemoteCollection` interface provides just the `size()` method and it extends the `Iterable` interface and not the `Collection` as the name would imply.

Reply:
I am going to submit a PR to have `StaticRemoteCollection<T>` implement `Collection<T>` and you can review it.

If we use `StaticRemoteCollection` as the return type here, we'll then get a `Collection` instead of an `Iterable` which makes sense since the whole point of that class is to store a collection.

Reply:
sure

Reply:
The PR is already in, so now you can change it to `isEmpty()` if you want. For `ArrayList`, there's no performance difference vs. `size()`, but it was a `Collection` already, so we might as well take advantage of it.

Reply:
nevermind, I would leave it as it is

The local `pncClient` returns RemoteCollection and changing it would require API change in PncClient interface which I like as it is.

Reply:
I see what you mean. We have no control over `RemoteCollection` which, despite its name, does not implement `Collection`. It also looks like doing so for `DefaultRemoteCollection` would be inefficient.